### PR TITLE
Detect spurious ; before assoc fn body

### LINKED ISSUE #105226: Detect spurious `;` before assoc fn body
When copying bounds from a trait assoc item, you can accidentally copy the semicolon. If you forget to remove it, *multiple* errors are emitted, while the second is:

```
error: associated function in `impl` without body
    --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:2349:5
     |
2349 | /     fn note_obligation_cause_code<T>(
2350 | |         &self,
2351 | |         err: &mut Diagnostic,
2352 | |         predicate: &T,
...    |
2357 | |     ) where
2358 | |         T: fmt::Display + ToPredicate<'tcx, T>;
     | |                                               ^ help: provide a definition for the function: `{ <body> }`
     | |_______________________________________________|
```

This should be explicitly handled to avoid knock down errors.

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #105226 

r? @TaKO8Ki

### CODE REVIEW

File: src/test/ui/parser/issues/issue-60075.stderr [hunk 1]
```diff
@@ -11,7 +11,10 @@ LL | trait T {
    |         - item list starts here
 ...
 LL |         });
-   |           ^ non-item starts here
+   |           ^
+   |           |
+   |           non-item starts here
+   |           help: consider removing this semicolon
```
Comment (TaKO8Ki) [inline src/test/ui/parser/issues/issue-60075.stderr:17]: In this case I think rustc should not emit this suggestion. The following comment I wrote explains that.

https://github.com/rust-lang/rust/blob/d43674e2c9b1a7b134ef76378c30e822c75c9aa1/compiler/rustc_parse/src/parser/item.rs#L711-L726

File: compiler/rustc_parse/src/parser/item.rs [hunk 2]
```diff
@@ -707,7 +707,7 @@ impl<'a> Parser<'a> {
             }
             match parse_item(self) {
                 Ok(None) => {
-                    let is_unnecessary_semicolon = !items.is_empty()
+                    let mut is_unnecessary_semicolon = !items.is_empty()
                         // When the close delim is `)` in a case like the following, `token.kind` is expected to be `token::CloseDelim(Delimiter::Parenthesis)`,
                         // but the actual `token.kind` is `token::CloseDelim(Delimiter::Bracket)`.
```
Comment (TaKO8Ki) [inline compiler/rustc_parse/src/parser/item.rs:712]:
```suggestion
                        // but the actual `token.kind` is `token::CloseDelim(Delimiter::Brace)`.
```

I mistakenly wrote `Bracket` for `Brace`. Could you fix that instead?

File: compiler/rustc_parse/src/parser/item.rs [hunk 3]
```diff
@@ -724,11 +724,14 @@ impl<'a> Parser<'a> {
                         //         Some(4)
                         //     }
                         // ```
-                        && self
-                            .span_to_snippet(self.prev_token.span)
-                            .map_or(false, |snippet| snippet == "}")
+                        && self.prev_token.kind == token::CloseDelim(Delimiter::Brace)
```
Comment (TaKO8Ki) [inline compiler/rustc_parse/src/parser/item.rs:727]: As commented below, I think we should use `span_to_snippet`.

File: compiler/rustc_parse/src/parser/item.rs (hunk 2 above)
Comment (chenyukang) [reply compiler/rustc_parse/src/parser/item.rs:712]: Ok，I didn't notice it, I will rollback my changes on your part.

I also think `TokenTreesReader::parse_token_tree` does a very weird thing for this scenario, I had some try to fix it:
https://github.com/rust-lang/rust/pull/104012#issuecomment-1311764832

currently pending on: https://github.com/rust-lang/rust/pull/104012

File: compiler/rustc_parse/src/parser/item.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_parse/src/parser/item.rs]: Updated!

File: src/test/ui/suggestions/issue-105226.stderr [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+LL | }
+   | - item list ends here
+
+error: associated function in `impl` without body
+  --> $DIR/issue-105226.rs:7:5
+   |
+LL |     fn hello<P>(&self, val: &P) where P: fmt::Display; {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
+   |                                                      |
+   |                                                      help: provide a definition for the function: `{ <body> }`
```
Comment (estebank) [inline src/test/ui/suggestions/issue-105226.stderr:20]: Ideally we would also avoid emitting this error by making the parser recover from the spurious ;

**Pull request merged**